packages/react/src/Image.tsx

```diff
@@ -0,0 +1,403 @@
+  // offset is just added to the initial width. Both values are represented in
+  // px.
+  const [initialWidth, setInitialWidth] = useState(0);
+  const [initialClientX, setInitialClientX] = useState(0);
+
+  // The editor's width in px.
+  const [editorWidth, setEditorWidth] = useState(
+    props.editor.domElement.firstElementChild!.clientWidth
+  );
+
+  // The image width, represented as a fraction of the editor's width.
+  const [width, setWidth] = useState(() => parseFloat(props.block.props.width));
+
+  // Creates an Uppy instance for file uploading.
+  // TODO: Server endpoints/URLs
+  const [uppy] = useState(() =>
+    new Uppy({ autoProceed: true, debug: true })
+      .use(Tus, {
+        endpoint: "https://master.tus.io/files/",
+      })
+      .use(GoogleDrive, {
+        companionUrl: "https://companion.uppy.io",
+      })
+      .use(Url, {
+        companionUrl: "https://companion.uppy.io",
+      })
+  );
+
+  // Takes a width value in px, converts it to a fraction of the editor's
+  // width, and updates the `width` state. Allows the image to be re-rendered
+  // without having to update the block.
+  const updateWidth = useMemo(
+    () => (newWidth: number) => {
+      newWidth = newWidth / editorWidth;
+
+      if (newWidth < minWidth) {
+        setWidth(minWidth);
+      } else if (newWidth > maxWidth) {
+        setWidth(maxWidth);
+      } else {
+        setWidth(newWidth);
+      }
+    },
+    [editorWidth]
+  );
+
+  // Sets up listeners for when the image is being resized.
+  useEffect(() => {
+    // Stops mouse movements from resizing the image and updates the block's
+    // `width` prop to the new value.
+    const mouseUpHandler = () => {
+      setResizeHandle(null);
+      props.editor.updateBlock(props.block, {
+        type: props.caption ? "captionedImage" : "image",
+        props: {
+          width: width.toString(),
+        },
+      });
+    };
+    // Re-renders the image with an updated width depending on the cursor
+    // offset from when the resize began, and which resize handle is being used.
+    const mouseMoveHandler = (e: MouseEvent) => {
+      if (!resizeHandle) {
+        return;
+      }
+
+      if (
+        textAlignmentToAlignItems(props.block.props.textAlignment) === "center"
+      ) {
+        if (resizeHandle === "left") {
+          updateWidth(initialWidth + (initialClientX - e.clientX) * 2);
+        } else {
+          updateWidth(initialWidth + (e.clientX - initialClientX) * 2);
+        }
+      } else {
+        if (resizeHandle === "left") {
+          updateWidth(initialWidth + initialClientX - e.clientX);
+        } else {
+          updateWidth(initialWidth + e.clientX - initialClientX);
+        }
+      }
+    };
+    // Re-renders the image when the viewport is resized. By storing the image
+    // width as a fraction of the editor's width, this allows the image to
+    // maintain its size relative to the editor.
+    const resizeHandler = () => {
```

Comment:
cool that resizing works! We're doing the entire implementation ourselves. Which is cool (and definitely helps to understand the inner workings). However, maybe it's worth investigating whether any hooks / libraries can help here. For example, https://react-grid-layout.github.io/react-resizable/index.html (not sure if it's the best one, it's the first one I saw). However, if the code of these libraries is super simple and similar to what we currently have, then maybe there's not a great reason to adopt them. On the other hand, if they identified and account for many edge-cases (I'm not sure), then maybe it's worthwhile

Reply:
Just tried it btw and resizing looks super nice!